File: Flutter/appreveal/lib/src/webview/webview_tools.dart
// Registers all 21 WebView MCP tools.

import 'dom_serializer.dart';
import 'webview_bridge.dart';
import '../mcp/mcp_router.dart';

void registerWebViewTools() {
  final router = MCPRouter.shared;
  final bridge = WebViewBridge.shared;

  // MARK: - get_webviews

  router.register(MCPToolDefinition(
    name: 'get_webviews',
    description: 'List all registered WebView instances with URL, title, and navigation state',
    inputSchema: {'type': 'object', 'properties': <String, dynamic>{}},
    handler: (_) async {
      final info = await bridge.webViewInfo();
      return {'webviews': info, 'count': info.length};
    },
  ));

  // MARK: - get_dom_tree

  router.register(MCPToolDefinition(
    name: 'get_dom_tree',
    description: 'Get the DOM tree of a web view. Returns full or partial DOM structure as JSON.',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
        'root': {'type': 'string', 'description': 'CSS selector for subtree root (default: body)'},
        'max_depth': {'type': 'integer', 'description': 'Max tree depth (default: 30)'},
        'visible_only': {'type': 'boolean', 'description': 'Only visible elements (default: false)'},
      },
    },
    handler: (params) async {
      final js = DOMSerializer.dumpTreeJS(
        root: params?['root'] as String?,
        maxDepth: params?['max_depth'] as int? ?? 30,
        visibleOnly: params?['visible_only'] as bool? ?? false,
      );
      try {
        final result = await bridge.evaluate(js: js, webViewId: params?['webview_id'] as String?);
        return {'dom': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - get_dom_interactive

  router.register(MCPToolDefinition(
    name: 'get_dom_interactive',
    description: 'Get all interactive DOM elements (inputs, buttons, links, selects) with their attributes, values, and selectors',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.interactiveJS(),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - query_dom

  router.register(MCPToolDefinition(
    name: 'query_dom',
    description: 'Query the DOM with a CSS selector. Returns matching elements with tag, text, attributes, rect.',
    inputSchema: {
      'type': 'object',
      'properties': {
        'selector': {'type': 'string', 'description': 'CSS selector'},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
        'limit': {'type': 'integer', 'description': 'Max results (default: 50)'},
      },
      'required': ['selector'],
    },
    handler: (params) async {
      final selector = params?['selector'] as String?;
      if (selector == null) return {'error': 'selector required'};
      final js = DOMSerializer.queryJS(selector, limit: params?['limit'] as int? ?? 50);
      try {
        final result = await bridge.evaluate(js: js, webViewId: params?['webview_id'] as String?);
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - find_dom_text

  router.register(MCPToolDefinition(
    name: 'find_dom_text',
    description: 'Find DOM elements containing specific text',
    inputSchema: {
      'type': 'object',
      'properties': {
        'text': {'type': 'string', 'description': 'Text to search for'},
        'tag': {'type': 'string', 'description': "Optional tag filter (e.g. 'button', 'a')"},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
      'required': ['text'],
    },
    handler: (params) async {
      final text = params?['text'] as String?;
      if (text == null) return {'error': 'text required'};
      final js = DOMSerializer.findTextJS(text, tag: params?['tag'] as String?);
      try {
        final result = await bridge.evaluate(js: js, webViewId: params?['webview_id'] as String?);
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - web_click

  router.register(MCPToolDefinition(
    name: 'web_click',
    description: 'Click a DOM element by CSS selector',
    inputSchema: {
      'type': 'object',
      'properties': {
        'selector': {'type': 'string', 'description': 'CSS selector'},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
      'required': ['selector'],
    },
    handler: (params) async {
      final selector = params?['selector'] as String?;
      if (selector == null) return {'error': 'selector required'};
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.clickJS(selector),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - web_type

  router.register(MCPToolDefinition(
    name: 'web_type',
    description: 'Type text into a DOM input or textarea by CSS selector',
    inputSchema: {
      'type': 'object',
      'properties': {
        'selector': {'type': 'string', 'description': 'CSS selector'},
        'text': {'type': 'string', 'description': 'Text to type'},
        'clear': {'type': 'boolean', 'description': 'Clear field first (default: false)'},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
      'required': ['selector', 'text'],
    },
    handler: (params) async {
      final selector = params?['selector'] as String?;
      final text = params?['text'] as String?;
      if (selector == null || text == null) return {'error': 'selector and text required'};
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.typeJS(selector, text, clear: params?['clear'] as bool? ?? false),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - web_select

  router.register(MCPToolDefinition(
    name: 'web_select',
    description: 'Select an option in a dropdown by CSS selector',
    inputSchema: {
      'type': 'object',
      'properties': {
        'selector': {'type': 'string', 'description': 'CSS selector for the select element'},
        'value': {'type': 'string', 'description': 'Option value to select'},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
      'required': ['selector', 'value'],
    },
    handler: (params) async {
      final selector = params?['selector'] as String?;
      final value = params?['value'] as String?;
      if (selector == null || value == null) return {'error': 'selector and value required'};
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.selectJS(selector, value),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - web_toggle

  router.register(MCPToolDefinition(
    name: 'web_toggle',
    description: 'Check or uncheck a checkbox/radio by CSS selector',
    inputSchema: {
      'type': 'object',
      'properties': {
        'selector': {'type': 'string', 'description': 'CSS selector'},
        'checked': {'type': 'boolean', 'description': 'Desired checked state'},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
      'required': ['selector', 'checked'],
    },
    handler: (params) async {
      final selector = params?['selector'] as String?;
      final checked = params?['checked'] as bool?;
      if (selector == null || checked == null) return {'error': 'selector and checked required'};
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.toggleJS(selector, checked),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - web_scroll_to

  router.register(MCPToolDefinition(
    name: 'web_scroll_to',
    description: 'Scroll a web view until a DOM element is visible',
    inputSchema: {
      'type': 'object',
      'properties': {
        'selector': {'type': 'string', 'description': 'CSS selector'},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
      'required': ['selector'],
    },
    handler: (params) async {
      final selector = params?['selector'] as String?;
      if (selector == null) return {'error': 'selector required'};
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.scrollToJS(selector),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - web_evaluate

  router.register(MCPToolDefinition(
    name: 'web_evaluate',
    description: 'Run arbitrary JavaScript in a web view and return the result',
    inputSchema: {
      'type': 'object',
      'properties': {
        'javascript': {'type': 'string', 'description': 'JavaScript to evaluate'},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
      'required': ['javascript'],
    },
    handler: (params) async {
      final js = params?['javascript'] as String?;
      if (js == null) return {'error': 'javascript required'};
      try {
        final result = await bridge.evaluate(js: js, webViewId: params?['webview_id'] as String?);
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - web_navigate

  router.register(MCPToolDefinition(
    name: 'web_navigate',
    description: 'Navigate a web view to a URL',
    inputSchema: {
      'type': 'object',
      'properties': {
        'url': {'type': 'string', 'description': 'URL to navigate to'},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
      'required': ['url'],
    },
    handler: (params) async {
      final url = params?['url'] as String?;
      if (url == null) return {'error': 'url required'};
      final controller = bridge.resolve(id: params?['webview_id'] as String?);
      if (controller == null) return {'error': 'WebView not found'};
      try {
        await controller.loadRequest(Uri.parse(url));
        return {'success': true, 'url': url};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - web_back

  router.register(MCPToolDefinition(
    name: 'web_back',
    description: 'Go back in web view history',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      final controller = bridge.resolve(id: params?['webview_id'] as String?);
      if (controller == null) return {'error': 'WebView not found'};
      if (!await controller.canGoBack()) return {'error': 'Cannot go back'};
      await controller.goBack();
      return {'success': true};
    },
  ));

  // MARK: - web_forward

  router.register(MCPToolDefinition(
    name: 'web_forward',
    description: 'Go forward in web view history',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      final controller = bridge.resolve(id: params?['webview_id'] as String?);
      if (controller == null) return {'error': 'WebView not found'};
      if (!await controller.canGoForward()) return {'error': 'Cannot go forward'};
      await controller.goForward();
      return {'success': true};
    },
  ));

  // MARK: - get_dom_links

  router.register(MCPToolDefinition(
    name: 'get_dom_links',
    description: 'Get all links on the page — just text and href. Minimal tokens.',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.linksJS(),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - get_dom_text

  router.register(MCPToolDefinition(
    name: 'get_dom_text',
    description: 'Get visible text content of the page stripped of all markup. Optionally scope to a CSS selector. Minimal tokens.',
    inputSchema: {
      'type': 'object',
      'properties': {
        'selector': {'type': 'string', 'description': 'CSS selector to scope text extraction (default: body)'},
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      final js = DOMSerializer.textContentJS(selector: params?['selector'] as String?);
      try {
        final result = await bridge.evaluate(js: js, webViewId: params?['webview_id'] as String?);
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - get_dom_forms

  router.register(MCPToolDefinition(
    name: 'get_dom_forms',
    description: 'Get all forms and their fields with types, names, values, options, and selectors. Includes pages without <form> tags.',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.formsJS(),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - get_dom_headings

  router.register(MCPToolDefinition(
    name: 'get_dom_headings',
    description: 'Get all headings (h1-h6) for page structure overview. Minimal tokens.',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.headingsJS(),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - get_dom_images

  router.register(MCPToolDefinition(
    name: 'get_dom_images',
    description: 'Get all visible images with src, alt text, and dimensions.',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.imagesJS(),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - get_dom_tables

  router.register(MCPToolDefinition(
    name: 'get_dom_tables',
    description: 'Get all tables with headers and row data.',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.tablesJS(),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));

  // MARK: - get_dom_summary

  router.register(MCPToolDefinition(
    name: 'get_dom_summary',
    description: 'Get a compact page summary: title, meta, headings (h1-h3), element counts (links, images, inputs, buttons), and form overview. Cheapest way to understand a page.',
    inputSchema: {
      'type': 'object',
      'properties': {
        'webview_id': {'type': 'string', 'description': 'Web view ID (default: first registered)'},
      },
    },
    handler: (params) async {
      try {
        final result = await bridge.evaluate(
          js: DOMSerializer.summaryJS(),
          webViewId: params?['webview_id'] as String?,
        );
        return {'result': result};
      } catch (e) {
        return {'error': e.toString()};
      }
    },
  ));
}

File: ReactNative/appreveal/android/src/main/kotlin/com/appreveal/webview/WebViewTools.kt
package com.appreveal.webview

import android.net.Uri
import android.webkit.WebView
import com.appreveal.mcpserver.MCPRouter
import com.appreveal.mcpserver.MCPToolDefinition
import com.appreveal.mcpserver.asIntOrNull
import com.appreveal.mcpserver.asStringOrNull
import com.appreveal.shared.MainThreadExecutor
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject

/**
 * Registers all 21 WebView MCP tools with the router,
 * matching the iOS WebViewTools.swift exactly.
 */
internal fun registerWebViewTools() {
    val router = MCPRouter

    // -- get_webviews --

    router.register(MCPToolDefinition(
        name = "get_webviews",
        description = "List all WebView instances on the current screen with URL, title, loading state, and frame",
        inputSchema = emptySchema(),
        handler = { _ ->
            val info = WebViewBridge.webViewInfo()
            val arr = JsonArray()
            for (wv in info) {
                arr.add(JsonObject().apply {
                    addProperty("id", wv["id"] as String)
                    addProperty("url", wv["url"] as String)
                    addProperty("title", wv["title"] as String)
                    addProperty("loading", wv["loading"] as Boolean)
                    addProperty("canGoBack", wv["canGoBack"] as Boolean)
                    addProperty("canGoForward", wv["canGoForward"] as Boolean)
                    addProperty("frame", wv["frame"] as String)
                })
            }
            JsonObject().apply {
                add("webviews", arr)
                addProperty("count", arr.size())
            }
        }
    ))

    // -- get_dom_tree --

    router.register(MCPToolDefinition(
        name = "get_dom_tree",
        description = "Get the DOM tree of a web view. Returns full or partial DOM structure as JSON.",
        inputSchema = webViewSchema(
            "root" to prop("string", "CSS selector for subtree root (default: body)"),
            "max_depth" to prop("integer", "Max tree depth (default: 30)"),
            "visible_only" to prop("boolean", "Only visible elements (default: false)")
        ),
        handler = { params ->
            val js = DOMSerializer.dumpTreeJS(
                root = params?.get("root")?.asStringOrNull(),
                maxDepth = params?.get("max_depth")?.asIntOrNull() ?: 30,
                visibleOnly = params?.get("visible_only")?.let {
                    if (it.isJsonPrimitive) it.asBoolean else false
                } ?: false
            )
            evalAndWrap(js, params?.get("webview_id")?.asStringOrNull(), "dom")
        }
    ))

    // -- get_dom_interactive --

    router.register(MCPToolDefinition(
        name = "get_dom_interactive",
        description = "Get all interactive DOM elements (inputs, buttons, links, selects) with their attributes, values, and selectors",
        inputSchema = webViewSchema(),
        handler = { params ->
            evalAndWrap(DOMSerializer.interactiveJS(), params?.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- query_dom --

    router.register(MCPToolDefinition(
        name = "query_dom",
        description = "Query the DOM with a CSS selector. Returns matching elements with tag, text, attributes, rect.",
        inputSchema = webViewSchema(
            "selector" to prop("string", "CSS selector"),
            "limit" to prop("integer", "Max results (default: 50)"),
            required = listOf("selector")
        ),
        handler = { params ->
            val selector = params?.get("selector")?.asString
                ?: return@MCPToolDefinition webViewError("selector required")
            val js = DOMSerializer.queryJS(selector, params.get("limit")?.asIntOrNull() ?: 50)
            evalAndWrap(js, params.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- find_dom_text --

    router.register(MCPToolDefinition(
        name = "find_dom_text",
        description = "Find DOM elements containing specific text",
        inputSchema = webViewSchema(
            "text" to prop("string", "Text to search for"),
            "tag" to prop("string", "Optional tag filter (e.g. 'button', 'a')"),
            required = listOf("text")
        ),
        handler = { params ->
            val text = params?.get("text")?.asString
                ?: return@MCPToolDefinition webViewError("text required")
            val js = DOMSerializer.findTextJS(text, params.get("tag")?.asStringOrNull())
            evalAndWrap(js, params.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- web_click --

    router.register(MCPToolDefinition(
        name = "web_click",
        description = "Click a DOM element by CSS selector",
        inputSchema = webViewSchema(
            "selector" to prop("string", "CSS selector"),
            required = listOf("selector")
        ),
        handler = { params ->
            val selector = params?.get("selector")?.asString
                ?: return@MCPToolDefinition webViewError("selector required")
            evalAndWrap(DOMSerializer.clickJS(selector), params.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- web_type --

    router.register(MCPToolDefinition(
        name = "web_type",
        description = "Type text into a DOM input or textarea by CSS selector",
        inputSchema = webViewSchema(
            "selector" to prop("string", "CSS selector"),
            "text" to prop("string", "Text to type"),
            "clear" to prop("boolean", "Clear field first (default: false)"),
            required = listOf("selector", "text")
        ),
        handler = { params ->
            val selector = params?.get("selector")?.asString
            val text = params?.get("text")?.asString
            if (selector == null || text == null) {
                return@MCPToolDefinition webViewError("selector and text required")
            }
            val clear = params.get("clear")?.let {
                if (it.isJsonPrimitive) it.asBoolean else false
            } ?: false
            evalAndWrap(DOMSerializer.typeJS(selector, text, clear), params.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- web_select --

    router.register(MCPToolDefinition(
        name = "web_select",
        description = "Select an option in a dropdown by CSS selector",
        inputSchema = webViewSchema(
            "selector" to prop("string", "CSS selector for the select element"),
            "value" to prop("string", "Option value to select"),
            required = listOf("selector", "value")
        ),
        handler = { params ->
            val selector = params?.get("selector")?.asString
            val value = params?.get("value")?.asString
            if (selector == null || value == null) {
                return@MCPToolDefinition webViewError("selector and value required")
            }
            evalAndWrap(DOMSerializer.selectJS(selector, value), params.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- web_toggle --

    router.register(MCPToolDefinition(
        name = "web_toggle",
        description = "Check or uncheck a checkbox/radio by CSS selector",
        inputSchema = webViewSchema(
            "selector" to prop("string", "CSS selector"),
            "checked" to prop("boolean", "Desired checked state"),
            required = listOf("selector", "checked")
        ),
        handler = { params ->
            val selector = params?.get("selector")?.asString
            val checked = params?.get("checked")?.let {
                if (it.isJsonPrimitive) it.asBoolean else null
            }
            if (selector == null || checked == null) {
                return@MCPToolDefinition webViewError("selector and checked required")
            }
            evalAndWrap(DOMSerializer.toggleJS(selector, checked), params.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- web_scroll_to --

    router.register(MCPToolDefinition(
        name = "web_scroll_to",
        description = "Scroll a web view until a DOM element is visible",
        inputSchema = webViewSchema(
            "selector" to prop("string", "CSS selector"),
            required = listOf("selector")
        ),
        handler = { params ->
            val selector = params?.get("selector")?.asString
                ?: return@MCPToolDefinition webViewError("selector required")
            evalAndWrap(DOMSerializer.scrollToJS(selector), params.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- web_evaluate --

    router.register(MCPToolDefinition(
        name = "web_evaluate",
        description = "Run arbitrary JavaScript in a web view and return the result",
        inputSchema = webViewSchema(
            "javascript" to prop("string", "JavaScript to evaluate"),
            required = listOf("javascript")
        ),
        handler = { params ->
            val js = params?.get("javascript")?.asString
                ?: return@MCPToolDefinition webViewError("javascript required")
            evalAndWrap(js, params.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- web_navigate --

    router.register(MCPToolDefinition(
        name = "web_navigate",
        description = "Navigate a web view to a URL",
        inputSchema = webViewSchema(
            "url" to prop("string", "URL to navigate to"),
            required = listOf("url")
        ),
        handler = { params ->
            val urlStr = params?.get("url")?.asString
                ?: return@MCPToolDefinition webViewError("Invalid URL")
            val webViewId = params.get("webview_id")?.asStringOrNull()
            try {
                MainThreadExecutor.runBlocking {
                    val webView = WebViewBridge.resolveWebView(webViewId)
                        ?: throw WebViewError.NotFound(webViewId ?: "default")
                    webView.loadUrl(urlStr)
                }
                JsonObject().apply {
                    addProperty("success", true)
                    addProperty("url", urlStr)
                }
            } catch (e: Exception) {
                webViewError(e.message ?: "Navigation failed")
            }
        }
    ))

    // -- web_back --

    router.register(MCPToolDefinition(
        name = "web_back",
        description = "Go back in web view history",
        inputSchema = webViewSchema(),
        handler = { params ->
            val webViewId = params?.get("webview_id")?.asStringOrNull()
            try {
                MainThreadExecutor.runBlocking {
                    val webView = WebViewBridge.resolveWebView(webViewId)
                        ?: throw WebViewError.NotFound(webViewId ?: "default")
                    if (!webView.canGoBack()) throw Exception("Cannot go back")
                    webView.goBack()
                }
                JsonObject().apply { addProperty("success", true) }
            } catch (e: Exception) {
                webViewError(e.message ?: "Cannot go back")
            }
        }
    ))

    // -- web_forward --

    router.register(MCPToolDefinition(
        name = "web_forward",
        description = "Go forward in web view history",
        inputSchema = webViewSchema(),
        handler = { params ->
            val webViewId = params?.get("webview_id")?.asStringOrNull()
            try {
                MainThreadExecutor.runBlocking {
                    val webView = WebViewBridge.resolveWebView(webViewId)
                        ?: throw WebViewError.NotFound(webViewId ?: "default")
                    if (!webView.canGoForward()) throw Exception("Cannot go forward")
                    webView.goForward()
                }
                JsonObject().apply { addProperty("success", true) }
            } catch (e: Exception) {
                webViewError(e.message ?: "Cannot go forward")
            }
        }
    ))

    // -- get_dom_links --

    router.register(MCPToolDefinition(
        name = "get_dom_links",
        description = "Get all links on the page -- just text and href. Minimal tokens.",
        inputSchema = webViewSchema(),
        handler = { params ->
            evalAndWrap(DOMSerializer.linksJS(), params?.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- get_dom_text --

    router.register(MCPToolDefinition(
        name = "get_dom_text",
        description = "Get visible text content of the page stripped of all markup. Optionally scope to a CSS selector. Minimal tokens.",
        inputSchema = webViewSchema(
            "selector" to prop("string", "CSS selector to scope text extraction (default: body)")
        ),
        handler = { params ->
            val js = DOMSerializer.textContentJS(params?.get("selector")?.asStringOrNull())
            evalAndWrap(js, params?.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- get_dom_forms --

    router.register(MCPToolDefinition(
        name = "get_dom_forms",
        description = "Get all forms and their fields with types, names, values, options, and selectors. Includes pages without <form> tags.",
        inputSchema = webViewSchema(),
        handler = { params ->
            evalAndWrap(DOMSerializer.formsJS(), params?.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- get_dom_headings --

    router.register(MCPToolDefinition(
        name = "get_dom_headings",
        description = "Get all headings (h1-h6) for page structure overview. Minimal tokens.",
        inputSchema = webViewSchema(),
        handler = { params ->
            evalAndWrap(DOMSerializer.headingsJS(), params?.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- get_dom_images --

    router.register(MCPToolDefinition(
        name = "get_dom_images",
        description = "Get all visible images with src, alt text, and dimensions.",
        inputSchema = webViewSchema(),
        handler = { params ->
            evalAndWrap(DOMSerializer.imagesJS(), params?.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- get_dom_tables --

    router.register(MCPToolDefinition(
        name = "get_dom_tables",
        description = "Get all tables with headers and row data.",
        inputSchema = webViewSchema(),
        handler = { params ->
            evalAndWrap(DOMSerializer.tablesJS(), params?.get("webview_id")?.asStringOrNull(), "result")
        }
    ))

    // -- get_dom_summary --

    router.register(MCPToolDefinition(
        name = "get_dom_summary",
        description = "Get a compact page summary: title, meta, headings (h1-h3), element counts (links, images, inputs, buttons), and form overview. Cheapest way to understand a page.",
        inputSchema = webViewSchema(),
        handler = { params ->
            evalAndWrap(DOMSerializer.summaryJS(), params?.get("webview_id")?.asStringOrNull(), "result")
        }
    ))
}

// -- Helper functions --

private fun evalAndWrap(js: String, webViewId: String?, resultKey: String): JsonElement {
    return try {
        val result = WebViewBridge.evaluate(js, webViewId)
        JsonObject().apply { addProperty(resultKey, result) }
    } catch (e: Exception) {
        webViewError(e.message ?: "Evaluation failed")
    }
}

private fun webViewError(message: String): JsonElement {
    return JsonObject().apply { addProperty("error", message) }
}

private fun prop(type: String, description: String? = null): JsonObject {
    return JsonObject().apply {
        addProperty("type", type)
        if (description != null) addProperty("description", description)
    }
}

private fun emptySchema(): JsonObject {
    return JsonObject().apply {
        addProperty("type", "object")
        add("properties", JsonObject())
    }
}

private fun webViewSchema(vararg props: Pair<String, JsonObject>, required: List<String>? = null): JsonObject {
    return JsonObject().apply {
        addProperty("type", "object")
        add("properties", JsonObject().apply {
            add("webview_id", prop("string", "Web view ID (default: first on screen)"))
            for ((name, schema) in props) {
                add(name, schema)
            }
        })
        if (required != null) {
            add("required", JsonArray().apply { required.forEach { add(it) } })
        }
    }
}
